### PR TITLE
Consider status code 429 as recoverable errors to avoid resharding

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

Fixes: #8418  

We consider 400 status code same as 200 status code for the purpose of allowing resharding. However, if the code is already rate-limiting (i.e.,429), this will make the matter worse. Hence, take this as recoverable errors and fall in the backoff loop. Also, 100 milliseconds is short for a 429, so maybe 1 second be a good balance.

cc @csmarchbanks @cstyan

### CODE REVIEW

**Comment by roidelapluie** [comment]
This is a breaking change.


I think this is a good idea, because it is very anniyoing in practice, but should be configurable.

File: storage/remote/client.go
Comment (roidelapluie) [inline storage/remote/client.go:203]: You don't need that complexity

```suggestion
	if httpResp.StatusCode == 429 {
```

**Comment by Harkishen-Singh** [comment]
@roidelapluie with configurable, do you mean to give users the ability to enable to disable backoff for rate-limiting?

**Comment by roidelapluie** [comment]
retryability, yes. (but I am not a remote write maintainer)

**Comment by csmarchbanks** [comment]
First, thanks for working on this, I think something like this behavior is certainly useful as dropping 429s indiscriminately is not ideal.

That said, I am not sure we ever decided what the best behavior on a 429 is. If you encounter a 429 in regular operation, chances are your remote write will just continue to fall behind until you force a restart dropping a significant chunk of data. Right now the 429 behavior can be useful because dropping some samples basically means less resolution in your data which might be better for many people not having any recent data at all.

Configuration of what to do on a 429 would be ok, but are there other ideas for handling rate limits more gracefully? Sending a new request that will probably fail 30ms later doesn't seem great. Limited number of retries + respect for the `Retry-After` header could be one option? This might end up relating to #7912.

**Comment by Harkishen-Singh** [comment]
I didn't get the retry after thing. Is it something that remote storage will respond to the remote write component, a time after which only, it should retry?

**Comment by csmarchbanks** [comment]
The `Retry-After` response header may be returned by remote storage to tell the client (remote write) how long to wait before trying another send. It is called out in https://tools.ietf.org/html/rfc6585#section-4.

To be clear, I am not saying implement the Retry-After logic. I am just not sure the 429 behavior should be the exact same as 5XX retry behavior, and would like a discussion. I am curious what @cstyan thinks too.

**Comment by Harkishen-Singh** [comment]
Ah, I feel like rate limiting is a thing from the remote storage. That means, the remote storage should get more control as to how it wants a particular request to behave. This gives itself the ability to handle the situation and come out of it. So, the response header should be a good way out here.

**Comment by cstyan** [comment]
If there's a standard for how to respond to 429s via the Retry-After header I think we should support it. I'm hesitant to include an option for `Max-Retries` given the work we've put into remote write to attempt to never lose data, but not strongly opposed.

**Comment by Harkishen-Singh** [comment]
Ah, any suggestions on how we should move ahead here?

**Comment by cstyan** [comment]
Supporting `Retry-After` when we see a 429 seems like the way to go to me, @csmarchbanks ?

**Comment by Harkishen-Singh** [comment]
@csmarchbanks @cstyan I have added the support for `Retry-After` and `Max-Retries` (since the discussion seems to be favourable towards some max-retries limit) as input from the response headers from remote storage systems. Can you please give towards the implementation?

File: storage/remote/client.go
Comment (csmarchbanks) [inline storage/remote/client.go:150]: I don't think `isRateLimiting` is necessary. A 0 value for `retryAfter` will mean to use the built in backoff instead.

File: storage/remote/client.go
Comment (csmarchbanks) [inline storage/remote/client.go:209]: https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.37 specifies that the `Retry-After` header can be either a duration in seconds, or a date time. If we support `Retry-After` we should support both of those forms, not the duration parsing that is done with `model.ParseDuration`.

**Review by csmarchbanks — COMMENTED** [review]
I am in favor of supporting the `Retry-After` header, but I don't think we should add `Max-Retries` as a header response or as part of this PR. Looking at https://github.com/prometheus/prometheus/pull/8237#issuecomment-739063187, Callum is also hesitant, so let's focus on just 429 behavior and `Retry-After` in this PR.

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:1088]: Somewhat a reminder for me in case a rebase doesn't go well, this should be `Warn` level, and failed should be Failed.

Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go]: I have implemented this in the recent push.

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go]: I have updated the logic to implement both the `seconds` way and the `date-time` one. I am not sure if the `date-time` one that I implemented is the best approach. So, a review would be helpful there as well.

File: storage/remote/client.go
Comment (csmarchbanks) [inline storage/remote/client.go:232]: Let's break this code out into a separate function that checks for the retry-after header and returns a duration (which would be 0 in the case of a missing header or error).

File: storage/remote/client.go
Comment (csmarchbanks) [inline storage/remote/client.go:219]: Rather than the `isDateTime` logic, you could just turn this into a `Duration` using `parsedTime.Sub(time.Now())`. Then logic elsewhere only has to care about handling a `Duration`.

File: storage/remote/queue_manager.go
Comment (csmarchbanks) [inline storage/remote/queue_manager.go:1102]: I think this could be a debug log. I would expect it to happen sometimes and it is a reasonable behavior. You could also get rid of it by always sleeping the larger amount of `backoff` or `retryAfter`.

**Review by csmarchbanks — COMMENTED** [review]
Some good progress, a couple ideas mostly to simplify the code!

File: storage/remote/client.go
Comment (csmarchbanks) [inline storage/remote/client.go:246]: You don't need to do both an `Atoi` and a `ParseDuration`. I would say either do
```
d, err := model.ParseDuration(t+"s")
// handle error
```
or
```
d, err := strconv.Atoi(t)
// handle error
return model.Duration(time.Duration(d) * time.Seconds)
```

File: storage/remote/client.go
Comment (csmarchbanks) [inline storage/remote/client.go:229]: I don't think we need to return an error here. All of the error handling behavior ends up the same as if the defaultBackoff is used. So in the case of an error just return the same as if there is no header, and simplify logic elsewhere.

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go:246]: oh yes, that was unnecessary

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go]: Done

File: config/config.go
Comment (csmarchbanks) [inline config/config.go:122]: Hmm, the default max backoff change is not really related to handling 429s better. Could that be part of a different PR if we want to do it?

File: storage/remote/client.go
Comment (csmarchbanks) [inline storage/remote/client.go:218]: I think http.TimeFormat would be the best format to use for the Parse call: https://pkg.go.dev/net/http#TimeFormat.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go:1072]: Yeah, but here we just aim with one incorrect result (`invalidRetryAfter`).

File: storage/remote/queue_manager.go
Comment (roidelapluie) [inline storage/remote/queue_manager.go:1080]: This could block rsharding for hours. I think we need a select that checks for context cancelation here.

File: storage/remote/client.go
Comment (roidelapluie) [inline storage/remote/client.go:209]:
```suggestion
	if httpResp.StatusCode == http.StatusTooManyRequests {
```

File: storage/remote/queue_manager.go
Comment (roidelapluie) [reply storage/remote/queue_manager.go:1072]: Why is that? A broken remote could send both -1 or -10 as header.

File: storage/remote/client.go
Comment (roidelapluie) [inline storage/remote/client.go:221]: If we check for < 0 later on we do not need this.

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go:209]: If we want to name via constants, then should other codes be changes similarly?

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go:221]: We can do this way as well or have a unit test for this function (which seems more reliable). For now, I am going with the later.

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go:231]: Thank you for catching this. Yes, we need to multiply by time.Second.

File: config/config.go
Comment (Harkishen-Singh) [reply config/config.go]: Sure, I have reverted this change.

File: storage/remote/queue_manager.go
Comment (roidelapluie) [inline storage/remote/queue_manager.go:1085]:
```suggestion
	select {
	case <-ctx.Done():
	case <-time.After(time.Duration(sleepDuration)):
}
```

Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go]: Thank you for this.

File: storage/remote/client.go
Comment (csmarchbanks) [reply storage/remote/client.go:221]: Yeah, the more I think about it the more I would prefer the check for < 0 rather than depending on -1. A whole lot more invalid cases are covered with that check in the case of future changes.

File: storage/remote/client.go
Comment (csmarchbanks) [inline storage/remote/client.go:148]: This probably doesn't need to be a constant since 0 is the default value anyway. I can see an argument for readability so I don't feel strongly.

**Review by csmarchbanks — APPROVED** [review]
:+1:, Thanks! one small comment but I am good with this as it is.

If no one else chimes in soon I will merge in time for the release this week.

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go:148]: Yeah, I feel that would be helpful for readability especially for `retryAfterDuration()` func.

**Comment by roidelapluie** [comment]
Is this a breaking change and do we want users to opt out?

**Comment by Harkishen-Singh** [comment]
I don't think existing users would break due to this (as we return default backoff in case the header is not set).

**Comment by csmarchbanks** [comment]
I think retrying 429s classifies as a breaking change if anyone is depending on them not being retried. It should certainly be listed as a [CHANGE].

I would be fine with a temporary flag to allow opting out for a couple of releases, but I am not sure it is required. Happy to defer to your judgement as the release shepherd :)

**Comment by roidelapluie** [comment]
> I don't think existing users would break due to this (as we return default backoff in case the header is not set).

Well it would break users who would not expect a retry at all.

**Comment by Harkishen-Singh** [comment]
@roidelapluie got it!

File: storage/remote/queue_manager.go
Comment (roidelapluie) [inline storage/remote/queue_manager.go:1072]:
```suggestion
			level.Debug(l).Log("msg", "retry-after cannot be in past, retrying using default backoff mechanism")
```

File: storage/remote/client.go
Comment (roidelapluie) [inline storage/remote/client.go:218]:
```suggestion
		return model.Duration(s * time.Second)
```

File: storage/remote/client.go
Comment (roidelapluie) [inline storage/remote/client.go:225]:
```suggestion
	return model.Duration(d*time.Second)
```

**Comment by roidelapluie** [comment]
LGTM after the last nits. I think that the logs would log that there is a 429, so that should be good for users to understand what is going on.

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go:218]: Sorry, this is a compilation error. We cannot multiply s (float64) and time.Second (time.Duration) and return to a model.Duration

File: storage/remote/client.go
Comment (Harkishen-Singh) [reply storage/remote/client.go:225]: Compilation error. d is int and time.Second is duration, and further, we want to return to model.Duration

**Comment by Harkishen-Singh** [comment]
@csmarchbanks , a note to check the logs as that is changed to log 429 if it is received.

File: storage/remote/queue_manager.go
Comment (Harkishen-Singh) [reply storage/remote/queue_manager.go]: Done

**Review by csmarchbanks — APPROVED** [review]
Thanks! :+1: from me. @roidelapluie would you take another look to make sure your comments have been addressed?

**Comment by roidelapluie** [comment]
LGTM, we could add later a flag to disable this retries if we see issues in the wild.

**Comment by gouthamve** [comment]
This is indeed a breaking change. From what I can we see, we _never_ drop data here? This means Prometheus will just keep falling behind if it is being constantly rate-limited? This assumes that Prometheus doesn't send data to a remote system that basically limits it constantly.

This is not true, we limit people in GrafanaCloud pretty regularly and they're completely fine with it. Further, in Cortex, we return 429 not just on a samples per second limit and do it for other cases too. For example, we have a limit for _active series_ and if a user sends more than their active series limit, we return a 429. In this case, Prometheus would _never_ be able to proceed because it would constantly try to send the sample that creates a new series and it would fail forever.

Now I am not sure if Cortex has the right behavior with 429 for other limits, it seemed pretty right when we built it ;) I still think it has the right behavior, but given Cortex has a major chunk of remote-write use-cases, we shouldn't roll this change out until we fix it there. WDYT @roidelapluie @csmarchbanks?

**Comment by roidelapluie** [comment]
My choice would be not to break users and threat 429 and non-recoverable by default, with the option to opt-in for this retries. That default could be changed in Prometheus 3.x, I do not think we should depend on only changes in remote writes to make that switch.

**Comment by Harkishen-Singh** [comment]
>  429 not just on a samples per second limit and do it for other cases too

@gouthamve I am sorry but isn't this very specific to cortex.

**Comment by pracucci** [comment]
> But, unless the Retry-After is specified, it would continue as earlier.

@Harkishen-Singh I don't think so. Before this PR, a 429 was considered a non recoverable error. After this PR, a 429 is always retried until succeed, either it contains the `Retry-After` or not.

**Comment by roidelapluie** [comment]
@Harkishen-Singh this is false, as if 

> > 429 not just on a samples per second limit and do it for other cases too
> 
> @gouthamve I am sorry but isn't this very specific to cortex?
> 
> That looks like a custom usage of 429 (I can be wrong). But, unless the `Retry-After` is specified, it would continue as earlier.

This is not exact: before, we would not retry at all. Now, with 429 and no retry-after, we would retry with the defaut backoff.

**Comment by roidelapluie** [comment]
Note: #8474 is open.

**Comment by ranton256** [comment]
Sorry, I think this change is only safe if you don't retry without the header, so it would default to the same behavior as previously for 429, and it should be behind a flag probably. Otherwise this will retry even without the header from the remote storage backend, which could be very problematic when 429's are returned for rate limiting or even reasons that are not ever going to succeed on retry.

**Comment by csmarchbanks** [comment]
We are putting it behind a flag for this release, and will likely put rules such as what you suggest or limited number of retries in future releases. See #8474 and #8477.

**Pull request merged**
